crates/ty_python_semantic/src/types/call/arguments.rs

```diff
@@ -108,6 +117,72 @@ impl<'a, 'db> CallArgumentTypes<'a, 'db> {
     pub(crate) fn iter(&self) -> impl Iterator<Item = (Argument<'a>, Type<'db>)> + '_ {
         self.arguments.iter().zip(self.types.iter().copied())
     }
+
+    /// Returns an iterator on performing [argument type expansion].
+    ///
+    /// Each element of the iterator represents a set of argument lists, where each argument list
+    /// contains the same arguments, but with one or more of the argument types expanded.
+    ///
+    /// [argument type expansion]: https://typing.python.org/en/latest/spec/overload.html#argument-type-expansion
+    pub(crate) fn expand(&self, db: &'db dyn Db) -> impl Iterator<Item = Vec<Vec<Type<'db>>>> + '_ {
+        /// Represents the state of the expansion process.
+        ///
+        /// This is useful to avoid cloning the initial types vector if none of the types can be
+        /// expanded.
+        enum State<'a, 'db> {
+            Initial(&'a Vec<Type<'db>>),
+            Expanded(Vec<Vec<Type<'db>>>),
+        }
+
+        impl<'db> State<'_, 'db> {
+            fn len(&self) -> usize {
+                match self {
+                    State::Initial(_) => 1,
+                    State::Expanded(expanded) => expanded.len(),
+                }
+            }
+
+            fn iter(&self) -> impl Iterator<Item = &Vec<Type<'db>>> + '_ {
+                match self {
+                    State::Initial(types) => Either::Left(std::iter::once(*types)),
+                    State::Expanded(expanded) => Either::Right(expanded.iter()),
```

Comment:
nit: `expanded` is a slice iterator, so I think you should be able to replace `std::iter::once` with `std::slice::from_ref` and not have to use `Either`.